### ISSUE TITLE
LightGBM on ML.NET trains slower than LightGBM command line

### ISSUE THREAD

**Issue by vivekpradhan** [dotnet/machinelearning#532]
### System information

.NET Core SDK (reflecting any global.json):
 Version:   2.1.301
 Commit:    59524873d6

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04

Hardware:
Google Cloud default 64 core machine.

### Issue

Ran training on same dataset with same params.
Dataset: 25k features x 140k rows (balanced binary classes)
Params: 
Command Line: 
```
./lightgbm metric=binary_logloss min_data_in_leaf=500 bagging_fraction=0.8 boosting_type=gbdt bagging_freq=5 max_bin=255 objective=binary valid_data=../../../pedata/test.csv max_depth=10 feature_fraction=0.8 num_leaves=70 output_result=prediction.txt num_machines=1 learning_rate=0.1 output_model=LightGBM_model.txt data=../../../pedata/train.csv num_threads=64 task=train is_training_metric=true num_iterations=500 metric_freq=1 tree_learner=serial
```
ML.NET
```
var lclassifier = new LightGbmBinaryClassifier() { UseCat=false, MaxBin= 255, EvalMetric= LightGbmArgumentsEvalMetricType.Logloss , NumLeaves= 70, NThread= 64,LearningRate= 0.1, NumBoostRound= 500, MinDataPerLeaf= 500};
            lclassifier.Booster = new GbdtBoosterParameterFunction() { 
                MaxDepth = 10,
                FeatureFraction=0.8,
                SubsampleFreq=5 };
            pipeline.Add(lclassifier);
```
- **What happened?**

Both use all the 64 cores (As seen on htop).
Data Loading + Training Time: Command Line: 10mins, ML.NET: 17mins

- **What did you expect?**
Since both were build on my machine from source, I expected that the training time would be comparable.  

Why is the ML.NET implementation slower? Is there something I can do to speed it up?

```
Stopwatch stopwatch = new Stopwatch();
stopwatch.Start();
// STEP 5: Train your model based on the data set
var model = pipeline.Train<IrisData, IrisPrediction>();
stopwatch.Stop();
```
The above stopwatch gives me time taken for Loading+Training. Is there a way to check the time taken only for the training step. Right now I am not sure if the data loading is slow or the training is slow.

-----------------
Edit: I got approximate training time by assuming that when CPU usage spikes, thats when training starts.
Training Time: Command Line: 335s, ML.NET:680s (approx)
Data Loading: Command Line: 265s, ML.NET: 340s (approx)

**Comment by Ivanidzo4ka** [comment]
@vivekpradhan  
>>Since both were build on my machine from source.
Just to be clear, in ML.NET we use nuget provided by LightGBM with prebuild dll.
Can you share how you prebuild lightgbm on Linux? was it MPI version, any specific options or just default cmake call?

**Comment by vivekpradhan** [comment]
I did this:
```
git clone --recursive https://github.com/Microsoft/LightGBM ; cd LightGBM
mkdir build ; cd build
cmake ..
make -j4
```
I saw instructions for this from here: https://lightgbm.readthedocs.io/en/latest/Installation-Guide.html
So this is not the MPI version, and I run it on a single machine.

**Comment by eerhardt** [comment]
@vivekpradhan - how did you build ML.NET?  Did you just type `./build.sh`?  If so, that's a "Debug" version of the ML.NET code, which could be the factor in why it is slower.

Can you try building it with `./build.sh -release`?  That will get you ML.NET binaries that will be faster than debug binaries.

**Comment by codemzs** [comment]
Hi @vivekpradhan ,

I did perf comparison between ML.NET lightgbm and command line using breast cancer dataset that we have under the test folder and the main difference that I see is in data loading but the training time is comparable. In ML.NET we load the dataset in memory and pin it and then pass the reference to native lightgbm code that does the training. I believe it is the same native lightgbm code that is also used by the command line executable. 

ML.NET:
Not adding a normalizer.
Auto-tuning parameters: LearningRate = 0.2
LightGBM objective=binary
Not training a calibrator because it is not needed.
[1] 'Loading data for LightGBM' started.
**[1] 'Loading data for LightGBM' finished in 00:00:00.1107676.**
[2] 'Training with LightGBM' started.
[2] (00:00.02)  Iteration: 50
[2] (00:00.03)  Iteration: 100
**[2] 'Training with LightGBM' finished in 00:00:00.0681632.**

Command line LightGBM:
E:\machinelearning\packages\lightgbm\2.1.2.2\runtimes\win-x64\native>lightgbm.exe metric=binary_logloss min_data_in_leaf=20 boosting_type=gbdt max_bin=255 objective=binary num_leaves=20 output_result=prediction.txt num_machines=1 output_model=LightGBM_model.txt data=breast-cancer.txt task=train num_iterations=100 metric_freq=1 tree_learner=serial
[LightGBM] [Info] Finished loading parameters
[LightGBM] [Info] Finished loading data in 0.028311 seconds
[LightGBM] [Info] Number of positive: 241, number of negative: 458
[LightGBM] [Info] Total Bins 97
[LightGBM] [Info] Number of data: 699, number of used features: 9
[LightGBM] [Info] Finished initializing training
[LightGBM] [Info] Started training...
**[LightGBM] [Info] 0.124142 seconds elapsed, finished iteration 100**
[LightGBM] [Info] Finished training

**Comment by vivekpradhan** [comment]
where are the logs generated for ML.NET?

**Comment by codemzs** [comment]
You will need to set verbose=true in the constructor of TlcEnvironment [here](https://github.com/dotnet/machinelearning/blob/0f94a3b12be82f476037d4a4ab6be06bbdcadd57/src/Microsoft.ML/LearningPipeline.cs#L164) and recompile the code. Then it will show the logs in the console when you train.

**Comment by codemzs** [comment]
Investigation is not yet over. I have asked for logs from Vivek for his experiments which he ran in the cloud to be sure there is an issue to investigate. 25k features with 140k rows could hit code paths that smaller datasets might not hit and vice-versa.

**Comment by vivekpradhan** [comment]
./lightgbm metric=binary_logloss min_data_in_leaf=500 bagging_fraction=0.8 boosting_type=gbdt bagging_freq=5 max_bin=255 objective=binary valid_data=../../pedata/test.csv max_depth=10 feature_fraction=0.8 num_leaves=70 output_result=prediction.txt num_machines=1 learning_rate=0.1 output_model=LightGBM_model.txt data=../../pedata/train.csv num_threads=64 task=train num_iterations=500
```
[LightGBM] [Info] Iteration:1, valid_1 binary_logloss : 0.619869
[LightGBM] [Info] 0.959032 seconds elapsed, finished iteration 1
.
.
.
[LightGBM] [Info] 368.210366 seconds elapsed, finished iteration 499
[LightGBM] [Info] Iteration:500, valid_1 binary_logloss : 0.0463751
[LightGBM] [Info] 369.149284 seconds elapsed, finished iteration 500
[LightGBM] [Info] Finished training


ML.NET
          LightGBMBinary: Loading data for LightGBM: Finished.
          LightGBMBinary: Loading data for LightGBM: Elapsed 00:05:51.7977232.
          LightGBMBinary: Training with LightGBM: Started.
          LightGBMBinary: Training with LightGBM: LightGBM objective=binary
          LightGBMBinary: Training with LightGBM: Finished.
          LightGBMBinary: Training with LightGBM: Elapsed 00:15:05.0409319.

```

**Comment by vivekpradhan** [comment]
I think the time elapsed for LightGBM includes the data loading time. So Command line: ~ 6 mins vs ML.NET: ~ 9 mins.

**Comment by vivekpradhan** [comment]
I was able to recreate this using synthetic data.
Python code used to create the synthetic dataset.
```
import numpy as np
import pandas as pd


header = ['fet'+str(x) for x in range(24500)]
np.random.seed(seed=102030)
synthtrain = np.random.randint(2, size=(140000, 24500))
#df = pd.DataFrame(synthtrain,columns=header)
#df.to_csv("synthtrain.csv", header = False, index = False)
synthtest = np.random.randint(2, size=(60000, 24500))
df = pd.DataFrame(synthtest,columns=header)
df.to_csv("synthtest.csv", header = False, index = False)
```

I ran the lightgbm command line as:
```
./lightgbm metric=binary_logloss min_data_in_leaf=500 bagging_fraction=0.8 boosting_type=gbdt bagging_freq=5 max_bin=255 objective=binary max_depth=10 feature_fraction=0.8 num_leaves=70 output_result=prediction.txt num_machines=1 learning_rate=0.1 output_model=LightGBM_model.txt data=../../pedata/synthtest.csv num_threads=64 task=train num_iterations=500
```
Output:
```
[LightGBM] [Info] Finished loading parameters
[LightGBM] [Info] Finished loading data in 90.367588 seconds
[LightGBM] [Info] Number of positive: 29842, number of negative: 30158
[LightGBM] [Info] Total Bins 48998
[LightGBM] [Info] Number of data: 60000, number of used features: 24499
[LightGBM] [Info] Finished initializing training
[LightGBM] [Info] Started training...
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.497367 -> initscore=-0.010533
[LightGBM] [Info] Start training from score -0.010533
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] 0.492581 seconds elapsed, finished iteration 1
.
.
.
[LightGBM] [Info] 193.451655 seconds elapsed, finished iteration 500
[LightGBM] [Info] Finished training
```
ML.NET
```
string dataPath = "/home/vpradhan/pedata/synthtest.csv";
pipeline.Add(new TextLoader(dataPath).CreateFrom<MyData>(separator: ','));
pipeline.Add(new ColumnConcatenator("Features", "NumericFeatures"));
var lclassifier = new LightGbmBinaryClassifier() { UseCat=false, MaxBin= 255, EvalMetric= LightGbmArgumentsEvalMetricType.Logloss , NumLeaves= 70, NThread= 64,LearningRate= 0.1, NumBoostRound= 500, MinDataPerLeaf= 500};
lclassifier.Booster = new GbdtBoosterParameterFunction() { 
    MaxDepth = 10,
    FeatureFraction=0.8,
    SubsampleFreq=5 };
pipeline.Add(lclassifier);
```
Output:
```
LightGBMBinary: Loading data for LightGBM: Finished.
          LightGBMBinary: Loading data for LightGBM: Elapsed 00:04:06.7554922.
          LightGBMBinary: Training with LightGBM: Started.
          LightGBMBinary: Training with LightGBM: LightGBM objective=binary
          LightGBMBinary: Training with LightGBM: Finished.
          LightGBMBinary: Training with LightGBM: Elapsed 00:06:15.2976411.
        TrainLightGBM: Training: Not training a calibrator because it is not needed.
        TrainLightGBM: Training: Finished.
        TrainLightGBM: Training: Elapsed 00:10:22.1267456.
```

So 6 mins 15 sec for ML.NET and ~ 3mins 15 sec for command line.

**Comment by codemzs** [comment]
@vivekpradhan Thank you for give code for the synthetic dataset. I will investigate this further and get back.

**Comment by vivekpradhan** [comment]
Hi I ran the model again after removing bagging_fraction from the command line LGBM it now runs in 4mins 24secs

```
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] 263.563869 seconds elapsed, finished iteration 499
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] 264.097724 seconds elapsed, finished iteration 500
[LightGBM] [Info] Finished training
```

**Comment by codemzs** [comment]
Can you please post the exact command you ran?

**Comment by vivekpradhan** [comment]
This is the command I used. 
```
./lightgbm metric=binary_logloss min_data_in_leaf=500 boosting_type=gbdt bagging_freq=5 max_bin=255 objective=binary max_depth=10 feature_fraction=0.8 num_leaves=70 output_result=prediction.txt num_machines=1 learning_rate=0.1 output_model=LightGBM_model.txt data=../../pedata/synthtest.csv num_threads=64 task=train num_iterations=500
```

When I compare this to my C# code shown above the only difference is UseCat=false in the C# code. This is added because I found that the default in C# is true but in the command line is false (categorical_feature = ""). Note that if I leave this out then C# runs slower.

**Comment by Ivanidzo4ka** [comment]
DRI RESPONSE:
@codemzs  can you update this thread with your findings or insights what was the issue, and is it still present or not?

**Comment by shauheen** [comment]
ping @singlis

**Comment by singlis** [comment]
Hi @vivekpradhan, 

Are you able to repro the issue with the latest version of ML.net? ML.net is currently using LightGBM 2.2.1.1 (which has likely updated since the original post). I have run some benchmarks using different datasets against the python LightGBM package on Ubuntu 16.04 (using the Azure Data Science VM).

The Data Science VM uses LightGBM version 2.1.2 which is comes just before 2.2.1.1. So its not exactly and apples-to-apples comparison for versions, but this is close.

Here is what I observed for the training time of ML.net and LightGBM, note this it training time and does not include load time:
Time is in seconds, num iterations =500, num leaves =256, learning rate = 0.1.

- Higgs 
    - ML.net: 129.18834
    - LightGBM: 154.059

- Airline:
    - ML.net:  579.737158
    - LightGBM: 1079.23

I am happy to re-run new tests or look into this issue further. 

Thanks, 

Scott

**Comment by singlis** [comment]
I am closing this issue due to no response, but @vivekpradhan feel free to re-open if needed.
